File: problems/039/039.f90
PROGRAM F039

    ! Copyright 2021 Melwyn Francis Carlo

    IMPLICIT NONE

    INTEGER :: P_MAX_SOLUTIONS = 0
    INTEGER :: N_MAX_SOLUTIONS = 0

    INTEGER :: B, N, P

    REAL    :: A, B2, P2

    DO P = 1000, 500, -1

        B = 1
        N = 0

        DO
            IF (B == INT(REAL(P) / 2.0)) EXIT

            P2 = REAL(P)
            B2 = REAL(B)

            A = ((P2 * P2) - (2.0 * P2 * B2)) / ((2.0 * P2) - (2.0 * B2))

            IF (A - REAL(FLOOR(A)) <= 0) N = N + 1

            B = B + 1

        END DO

        IF (N > N_MAX_SOLUTIONS) THEN
            N_MAX_SOLUTIONS = N
            P_MAX_SOLUTIONS = P
        END IF

    END DO

    PRINT ('(I0)'), P_MAX_SOLUTIONS

END PROGRAM F039

File: problems/039/039.c
#include <stdio.h>

/* Copyright 2021 Melwyn Francis Carlo */

int main()
{
    int p_max_solutions = 0;
    int n_max_solutions = 0;

    for (int p = 1000; p > 500; p--)
    {
        int b = 0;
        int n = 0;
        while (1)
        {
            if (++b == (p / 2))
                break;

            float p2 = (float)p;
            float b2 = (float)b;

            float a = ((p2 * p2) - (2 * p2 * b2)) / ((2 * p2) - (2 * b2));

            if ((int)(a) == a)
                n++;
        }

        if (n > n_max_solutions)
        {
            n_max_solutions = n;
            p_max_solutions = p;
        }
    }

    printf("%d\n", p_max_solutions);

    return 0;
}
